### DISCUSSION TITLE
On capture statement and subspecies

### DISCUSSION THREAD

**Discussion by Aya-Badawy89** [gama-platform/gama#96] [Q&A] [answered]
I created the following dummy model
```
species negotiators control:fsm {
	
	species negotiating_stakeholders parent:stakeholders { }
}

species stakeholders control:fsm {
	int n <- 1;
	
	state initial_state initial:true {
		write 'I am in initial state';
		loop while: (n<5) {n <- n+1;}
		transition to:second_state when: n=5;
	}
	
	state second_state {
		write members;
		write 'I am in second state';	
		if (length(stakeholders) > 0) {
			ask negotiators {capture myself as:negotiating_stakeholders;}
		}				
	}
}
```
My question is: after I capture the agent to another species, the state (second state) is still running although technically there are no agents there to run it! The species stakeholders is empty, so I think it should stop.
Another question: can agents migrate between species with different control architectures?

**Comment by benoitgaudou** [comment, marked as answer]
Dear,
sorry for the delay in answering.

"My question is: after I capture the agent to another species, the state
(second state) is still running although technically there are no agents
there to run it! The species stakeholders is empty, so I think it should
stop."

For this question, the point is that your negotiating_stakeholders are kind
of stakeholders, inheriting all the behaviors (reflexes, state ...),
actions, attributes from the stakeholders. As a consequence, even when
captured they still execute the states of stakeholders?

If you want to avoid that:
- either you express that your negotiating_stakeholders are not scheduled
(i.e. executed) with the schedules facet

	species negotiating_stakeholders parent:stakeholders schedules: [] { }

- or you redefine the behavior of stakeholders in negotiating_stakeholders:

*species* negotiators *control:**fsm* {

*species* negotiating_stakeholders *parent:**stakeholders* /*schedules: []*/
{

*state* initial_state *initial*:true {

*write* 'Captured initial state ' *+* *self*;

}

*state* second_state {

*write* 'Captured second state ' *+* *self*;

}

}

}

For your second question,
"Another question: can agents migrate between species with different
control architectures?"
Technically they can. Just try to add 'control:simple_bdi' to the
negotiating_stakeholdders species.
Nevertheless the addition of this new architecture will drop the previous
one (state will not be executed any more).

Cheers

Benoti


Le ven. 10 sept. 2021 à 16:34, Aya-Badawy89 ***@***.***> a
écrit :

**Comment by Aya-Badawy89** [comment]
Thank you very much, Benoit
